### ISSUE TITLE
Prepare a TensorBoard pull request so other people can use this

### ISSUE THREAD

**Issue by chrisranderson** [chrisranderson/beholder#4]
@jart @dandelionmane:

I'll summarize the changes I think need to be made to TensorBoard here.

1. `tensorboard/BUILD`: 
  a. add dependency for the plugin. This seems problematic - what if I want to change the name of the plugin? Could you add `tensorboard/plugins/third-party` and then I put all of my stuff in there, and the following steps are automatically done? This could be a big request, but that way people can install new plugins just by cloning the repo, no TensorBoard update required. Maybe the third-party repos have some configuration file that TensorBoard reads for all the external changes that need to be made.

2. `tensorboard/components/tf_globals/globals.ts`: 
  a. add tab. In line with my thoughts in 1, I'm picturing instead of 

![image](https://user-images.githubusercontent.com/5461398/27975842-411d14b6-6321-11e7-8329-5cf2c6a51db3.png)

it is

![image](https://user-images.githubusercontent.com/5461398/27975876-5476fdc4-6321-11e7-8542-65a5a0725ec3.png)

with all of the third party plugins in a dropdown or something.

3. `tensorboard/components/tf_tensorboard/BUILD`
  a. add dependency for a new dashboard.

4. `tensorboard/components/tf_tensorboard/tf-tensorboard.html`:
  a. import new dashboard
  b. add template to content div
  c. add `_modeIs<plugin>` function

5. `tensorboard/main.py`
  a. add plugin to the list of plugins.

What do you think? Does this look right? I think these are all of the changes I've made to TensorBoard. I finally have the full pipeline working though! This is by no means final, but these frames are changing (requested at 10 FPS, just changing the source on an img tag using `setInterval`, which I think I'll change so that it automatically adjusts depending on whether it's receiving the same frame over and over or not):

![image](https://user-images.githubusercontent.com/5461398/27976311-1eb0f17a-6323-11e7-80b6-c1659a85408b.png)

**Comment by jart** [comment]
We can actually make it work out of the box from within this codebase. However I'm noticing that this repository doesn't have the web_library_example structure anymore. What I'll do asap is create a custom_tensorboard_example repo that shows how to do this.

**Comment by chrisranderson** [comment]
@jart oh, I just started following the pattern of plugins in TensorBoard. I never realized that I needed to structure it like web_library_example. Do I need to change it to match? How does it work with TensorBoard in that format?

**Comment by dandelionmane** [comment]
Tagging @wchargin too.

**Comment by jart** [comment]
@chrisranderson You're going to want both. It'll make more sense when I created the custom_tensorboard_example repository. See also: https://github.com/tensorflow/tensorboard/pull/206

**Comment by wchargin** [comment]
Hi—nice work!

To your points (1) and (5): instead of changing TensorBoard code, it
would be easiest for you to make a (verbatim) copy of TensorBoard's
`main.py` and add your plugin to the `plugins` list in the main
function. Then, likewise duplicate the Bazel rule and add your plugin as
a build dependency there. This should be the only place that the list of
plugins appears in TensorBoard, so it should be all that you need to
change. (I believe that this is what @jart is referring to in her
comment.)

(The only caveat of this is that some of the build dependencies of
`//tensorboard` have internal visibility, so I don't think that you'll
be able to view them. @jart: ideas?)

To your point about putting third-party plugins in a dropdown menu:
let's hold off on that for now, for two reasons. First, it's
implementation work. (For reference, I recently added a different
dropdown to that tabs pane, and…the tab bar and dropdown components
really don't interact in the way we want. Like, they work, but they
weren't made for each other: the styles are bad, the focus state isn't
shared, etc.) Second, it's not clear to me that this is the final state
that we want. We're trying to make our first-party plugins the least
privileged that they can be: that is, that the only difference between
first- and third-party plugins is the directory in which they're stored.
In the End Game, a TensorBoard user will be able to either build or
launch TensorBoard and select exactly the plugins they want, and those
will be the ones that they get. At that point, the
first-party/third-party distinction isn't relevant.

To your points (2), (3), and (4): it used to be the case that this was required, but over the past few weeks we've been simplifying the system and reducing the number of integration points. Once my open PR #202 lands, you will only have to make one change: add an entry
```js
{
    name: 'beholder',
    component: 'tf-beholder-dashboard',
    reloadEnabled: true,
},
```
to our `DASHBOARDS` manifest in `components/tf_tensorboard/dashboards.ts`.
(At the time of writing that file does not exist in master; check the
files of #202 to see what's going on.)

(This PR is approved, but our build is currently failing, so I'll merge
it on Monday after we fix the build.)

Again, you'll want to copy `dashboards.ts` and build a TensorBoard
against that. This is a bit different than copying `main.py`, because
it's a _dependency_ of the entry point instead of the entry point
itself, so you'll need to build the entry point as well. Ideally we
should have a Skylark macro for this—again, @jart probably knows what
the best thing to do is.

If you pull latest TensorBoard source before then, you'll find that the
steps are slightly similar in a different way. Given that the steps that
you identified above are correct as of a few weeks ago, I'm sure that
you'll figure out what the right things to do are. :-)

Best,
WC

**Comment by chrisranderson** [comment]
Heh. With fun, frequent updates come moving targets. :) So really at this point, maybe I wait a bit til @jart's new example is up? And then I can revert all of my changes in my copy of TensorBoard and start really parsing through @wchargin's comment (thanks for that!), which I'm guessing will largely if not totally be covered by `custom_tensorboard_example`.

Thank you all for your help! This has been a great experience.

**Comment by wchargin** [comment]
> wait a bit til @jart's new example is up

Yes; I'd advise that.

I'm assuming from your screenshots that you're able to work on a locally patched copy of TensorBoard and so this isn't blocking anything—right? If this isn't the case or stops being the case, do let us know and we can move up our timeframes.

Thanks!

**Comment by wchargin** [comment]
Sorry about the moving targets…but I promise that they're a Good Thing and will make your code Just Work! :-)

**Comment by chrisranderson** [comment]
@wchargin yep, I made the changes and used git diff for my first comment on this issue. :)

**Comment by chrisranderson** [comment]
@jart Any update on your custom_tensorboard_example repo? I'm about to the point where I'd like to ask other people to test this out on their own networks.

@wchargin I'm trying to figure out what it would take for others to install this plugin. It looks like I need to:
- at some point update my copy of TensorBoard. After https://github.com/tensorflow/tensorboard/pull/206 is merged?
- copy `main.py` into my repo, and alter it so that my plugin is listed. 
  - Somehow keep track of your version of `main.py`, and update mine whenever you change yours. Can this be avoided?
- copy the bazel rule and add my plugin to the dependencies.
  - Keep track of your version of the build rule, and update mine whenever you change yours. Can this be avoided?
- ~something with `dashboards.ts`~ something else, given https://github.com/tensorflow/tensorboard/pull/206?
- Run `bazel build tensorboard/plugins/beholder/tensorboard`
  - Since each plugin requires a rebuilt TensorBoard, rebuild every time you want to use a different third party plugin. Can this be avoided?

So far it looks like the installation steps for other people who want to use the plugin will be:

- clone the repo
- run a yet-to-be-written install script that will:
  - install Bazel if it isn't installed yet
  - rebuild TensorBoard using my custom build rule

Does that all sound right? Thanks for your help in this whole process.

**Comment by dandelionmane** [comment]
I think this is revealing a lot of areas of improvement for the plugin infrastructure - in particular, requiring bazel to install a new plugin locally is something we should try to avoid (and move towards self-contained Python modules that have everything they need? These python modules could be built with our bazel rules).

Re "copy main.py, and update whenever we change ours" - I think we could do this by factoring out a STANDARD_PLUGINS list and having our main.py use that list. Then people maintaining forks of main still get the default standard_plugins from upstream. (This would probably be a short term solution, we could make something better later.)

**Comment by jart** [comment]
I'm working on getting https://github.com/tensorflow/tensorboard/pull/206 submitted right now. The example repo comes next.

**Comment by chrisranderson** [comment]
@dandelionmane @wchargin If I understand right, they'd also have to clone the TensorBoard repo because the one that comes by pip installing TensorFlow doesn't have all the files necessary for building from scratch.

**Comment by wchargin** [comment]
> It looks like I need to:
>   - at some point update my copy of TensorBoard. After tensorflow/tensorboard#206 is merged?

Certainly. (It's merged.)

>   - copy main.py into my repo, and alter it so that my plugin is listed.
>   - copy the bazel rule and add my plugin to the dependencies.
>   - Keep track of your version of the build rule, and update mine whenever you change yours. Can this be avoided?
>   - Run bazel build tensorboard/plugins/beholder/tensorboard

Right; all this is true in the current model. Eventually, we want you
job to be as simple as writing a small constant number of build rules,
like (API not final):

```
# in //beholder/BUILD

# Normal Python build rule that defines your plugin.
py_library(
    name = "beholder_plugin",
    srcs = ["beholder_plugin.py"],
    deps = ["//tensorboard/plugins:base_plugin"],
)

# Special TensorBoard build rule that lets us hook your plugin into
# a TensorBoard application.
tensorboard_plugin(
    name = "beholder",
    main = "chrisranderson.beholder_plugin.BeholderPlugin",
    deps = [":beholder_plugin"],
)

# TensorBoard application that bundles all the plugins that you care
# about, in addition to the standard plugins.
tensorboard_binary(
  name = "tensorboard",
  deps = [
    ":beholder",
  ],
)

# $ bazel run //beholder:tensorboard
```
(with a corresponding way to link up the HTML frontend; also not
finalized).

>   - ~~something with dashboards.ts~~ something else, given tensorflow/tensorboard#206?

Add it here: https://github.com/tensorflow/tensorboard/commit/a990dac3a3c099e8702bccd2a03734d6173188cf#diff-67383bbdebb2989a5d45ae3c4a99d933R30

> Since each plugin requires a rebuilt TensorBoard, rebuild every time you want to use a different third party plugin. Can this be avoided?

This is a reasonable request, and indeed there are advantages to having
runtime plugin dependencies. (For instance, in the Java world many
applications that support plugins let you just drop a jarfile onto the
GUI and it'll dynamically find and load the plugins. Something like this
would be ncie.) There are also downsides, in that core TensorBoard can
make many fewer assumptions (there's no constant list of available
plugins). Furthermore, there's trickiness with implementation:
currently, all of our frontend components are packaged into one giant
HTML file at build time, and it's not clear how we would change this to
accommodate runtime plugin dependencies.

So, I'd file this as "nice-to-have, but not on our near-term radar."
I'd much rather you have an easy build-time solution than a clunky or
fragile runtime solution.

> So far it looks like the installation steps for other people who want to use the plugin will be:
>   - clone the tensorboard repo
>   - clone the plugin repo into tensorboard/plugins
>   - run a yet-to-be-written install script that will:
>       - install Bazel if it isn't installed yet
>       - rebuild TensorBoard using my custom build rule

See `tensorboard_binary` above. People should be able to just write such
a rule that pulls in the plugin dependencies automatically.

> Thanks for your help in this whole process.

Thank _you_ for your work on this plugin!

**Comment by chrisranderson** [comment]
@jart, do you know why the code as it is in this branch https://github.com/chrisranderson/beholder/tree/new-install/beholder doesn't show the new dashboard? It would also be nice to know why Bazel doesn't detect changes in e.g. `beholder/beholder_dashboard/beholder-dashboard.html`.

**Comment by chrisranderson** [comment]
@jart Nevermind, got it figured out. I wasn't using your `main.py` and my build rules weren't the same I don't think, so it was using the normal TensorBoard dashboard and not the one in my repo.

**Comment by chrisranderson** [comment]
Now that https://github.com/tensorflow/tensorboard/pull/206 is finished, the repo is in a pretty good state, and people can start trying it out. Thanks @jart!

**Comment by dandelionmane** [comment]
Woot! BTW Chris, you may want to take a look at the new tensorflow/tensorboard-plugin-example repo. Once this pull (https://github.com/tensorflow/tensorboard-plugin-example/pull/1) goes in, it will be a fully-fledged demo of creating a TB plugin. I'd appreciate any feedback on it (you can comment on the PR even if it's already merged when you look at it)
